src/coreclr/tools/Common/JitInterface/AsyncMethodDesc.cs

```diff
@@ -0,0 +1,142 @@
+                var ret = wrappedSignature.ReturnType;
+                var md = (MetadataType)ret;
+                Debug.Assert(md.Namespace.SequenceEqual("System.Threading.Tasks"u8));
+                ReadOnlySpan<byte> name = md.Name;
+                TypeDesc returnType = null;
+                if (name.SequenceEqual("Task"u8) || name.SequenceEqual("ValueTask"u8))
+                {
+                    returnType = this.Context.GetWellKnownType(WellKnownType.Void);
+                }
+                else if (name.SequenceEqual("Task`1"u8) || name.SequenceEqual("ValueTask`1"u8))
+                {
+                    Debug.Assert(md.HasInstantiation);
+                    returnType = md.Instantiation[0];
+                }
+                else
+                {
+                    throw new UnreachableException("AsyncMethodDesc should not wrap a non-Task-like-returning method");
+                }
+
+                var builder = new MethodSignatureBuilder(_wrappedMethod.Signature);
+                builder.ReturnType = returnType;
+                builder.Flags |= MethodSignatureFlags.AsyncCallConv;
+                return builder.ToSignature();
+            }
+        }
+
+#if !SUPPORT_JIT
+        // Same pattern as UnboxingMethodDesc: these should not escape JIT hashing scope.
+        protected override int ClassCode => throw new NotImplementedException();
+        protected override int CompareToImpl(MethodDesc other, TypeSystemComparer comparer) => throw new NotImplementedException();
+        protected override int ComputeHashCode() => _jitVisibleHashCode;
+        int IJitHashableOnly.GetJitVisibleHashCode() => _jitVisibleHashCode;
+#else
+        int IJitHashableOnly.GetJitVisibleHashCode() => _jitVisibleHashCode;
+#endif
+    }
+
+    internal static class AsyncMethodDescExtensions
+    {
+        /// <summary>
+        /// Returns the other async variant. If the supplied method is an async-callconv wrapper, returns the wrapped (Task-returning) method.
+        /// If it is a Task/ValueTask returning method, returns (and possibly creates) the async-callconv variant via the factory; otherwise null.
+        /// </summary>
+        public static MethodDesc GetOtherAsyncMethod(this MethodDesc method, AsyncMethodDescFactory factory)
+        {
+            if (method is null) return null;
+            if (method is AsyncMethodDesc amd)
+                return amd.Target; // unwrap
+
+            if (method.IsAsync && ReturnsTaskLike(method))
+                return factory.GetAsyncMethod(method);
+
+            return null;
+        }
+
+        public static bool ReturnsTaskLike(this MethodDesc method)
+        {
+            TypeDesc ret = method.GetTypicalMethodDefinition().Signature.ReturnType;
+            if (ret == null) return false;
+
+            if (ret is MetadataType md)
+            {
+                if (md.Namespace.SequenceEqual("System.Threading.Tasks"u8))
```

Comment:
This should check for system module too so this match does not match some random local Task definition.